File: samples/fsharp/getting-started/Clustering_CustomerSegmentation/CustomerSegmentation.Predict/ClusteringModelScorer.fs
namespace CustomerSegmentation.Model

open Microsoft.ML
open Microsoft.ML.Core.Data
open Microsoft.ML.Runtime.Api
open Microsoft.ML.Runtime.Data
open OxyPlot
open OxyPlot.Series
open System
open System.Collections.Generic
open System.IO
open System.Diagnostics
open System.Linq
open CustomerSegmentation.DataStructures
open Common.ConsoleHelper


type ClusteringModelScorer (mlContext: MLContext, pivotDataLocation : string, plotLocation : string, csvlocation : string, modelPath : string) =

    let loadModelFromZipFile = 
        use stream = new FileStream (modelPath, FileMode.Open, FileAccess.Read, FileShare.Read)
        mlContext.Model.Load stream

    let SaveCustomerSegmentationCSV (predictions : IEnumerable<ClusteringPrediction>) (csvlocation: string) = 
        consoleWriteHeader "CSV Customer Segmentation"
        use w = new StreamWriter(csvlocation)
        w.WriteLine("LastName,SelectedClusterId")
        w.Flush()
        predictions |> Seq.iter (fun p -> w.WriteLine(sprintf "%s,%s" p.LastName (p.SelectedClusterId.ToString())  ); w.Flush())

        printfn "CSV Location %s" csvlocation

    let SaveCustomerSegmentationPlotChart (predictions : IEnumerable<ClusteringPrediction>) (plotLocation : string) = 
        consoleWriteHeader "Plot Customer Segmentation"
        let plot = new PlotModel(
                        Title = "Customer Segmentation",
                        IsLegendVisible = true
                    )
        let clusters = predictions |> Seq.map (fun p -> p.SelectedClusterId) |> Seq.distinct |> Seq.sort

        clusters 
            |> Seq.iter (fun c -> 
                    let scatter = new ScatterSeries(MarkerType = MarkerType.Circle, MarkerStrokeThickness = 2., Title = sprintf "Cluster: %i" c, RenderInLegend = true)
                    let series = predictions |> Seq.where (fun p -> p.SelectedClusterId = c) |> Seq.map (fun p -> new ScatterPoint(p.Location.[0] |> float ,p.Location.[1] |> float)) |> Seq.toArray
                    scatter.Points.AddRange series
                    plot.Series.Add scatter
                )
        plot.DefaultColors <- OxyPalettes.HueDistinct(plot.Series.Count).Colors

        let exporter = new SvgExporter(Width = 600., Height = 400.)
        let fs = new FileStream(plotLocation, FileMode.Create)
        exporter.Export(plot, fs)
        printfn "Plot Loation: %s" plotLocation

    let OpenChartInDefaultWindow (plotLocation: string) = 
        printfn "Showing Chart..."
        let p = new Process()
        let psi = new ProcessStartInfo(plotLocation)
        psi.UseShellExecute <- true
        p.StartInfo <- psi
        p.Start() |> ignore
        ()


    member __.CreateCustomerClusters () =
        let reader = 
            TextLoader(mlContext, TextLoader.Arguments (
                                Column = [|
                                    TextLoader.Column("Features", DataKind.R4 |> Nullable, [|TextLoader.Range(0,31 |> Nullable)|])
                                    TextLoader.Column("LastName", DataKind.Text |> Nullable, 32)
                                |],
                                HasHeader = true,
                                Separator = ","
                    ))

        let data : IDataView = reader.Read(MultiFileSource (pivotDataLocation) :> IMultiStreamSource)

        //Apply data transformation to create prediction / clustering
        let predictions = loadModelFromZipFile.Transform(data).AsEnumerable<ClusteringPrediction>(mlContext, false).ToArray()

        //Generate data files with customer data grouped by clusters
        SaveCustomerSegmentationCSV predictions csvlocation

        //Plot/paint the clusters in a chart and open it with the by-default image-tool in Windows
        SaveCustomerSegmentationPlotChart predictions plotLocation
        OpenChartInDefaultWindow plotLocation
        

File: samples/csharp/getting-started/Clustering_CustomerSegmentation/CustomerSegmentation.Predict/ClusteringModelScorer.cs
using OxyPlot;
using OxyPlot.Series;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

using Common;
using CustomerSegmentation.DataStructures;
using Microsoft.ML;
using Microsoft.ML.Data;

namespace CustomerSegmentation.Model
{
    public class ClusteringModelScorer
    {
        private readonly string _pivotDataLocation;

        private readonly string _plotLocation;
        private readonly string _csvlocation;
        private readonly MLContext _mlContext;
        private ITransformer _trainedModel;

        public ClusteringModelScorer(MLContext mlContext, string pivotDataLocation, string plotLocation, string csvlocation)
        {
            _pivotDataLocation = pivotDataLocation;
            _plotLocation = plotLocation;
            _csvlocation = csvlocation;
            _mlContext = mlContext;
        }

        public ITransformer LoadModel(string modelPath)
        {
            _trainedModel = _mlContext.Model.Load(modelPath, out var modelInputSchema);
            return _trainedModel;
        }

        public void CreateCustomerClusters()
        {
            var data = _mlContext.Data.LoadFromTextFile(path:_pivotDataLocation,
                            columns: new[]
                                        {
                                          new TextLoader.Column("Features", DataKind.Single, new[] {new TextLoader.Range(0, 31) }),
                                          new TextLoader.Column(nameof(PivotData.LastName), DataKind.String, 32)
                                        },
                            hasHeader: true,
                            separatorChar: ',');
            
            //Apply data transformation to create predictions/clustering
            var tranfomedDataView = _trainedModel.Transform(data);
            var predictions = _mlContext.Data.CreateEnumerable <ClusteringPrediction>(tranfomedDataView, false)
                            .ToArray();

            //Generate data files with customer data grouped by clusters
            SaveCustomerSegmentationCSV(predictions, _csvlocation);

            //Plot/paint the clusters in a chart and open it with the by-default image-tool in Windows
            SaveCustomerSegmentationPlotChart(predictions, _plotLocation);
            OpenChartInDefaultWindow(_plotLocation);
        }

        private static void SaveCustomerSegmentationCSV(IEnumerable<ClusteringPrediction> predictions, string csvlocation)
        {
            ConsoleHelper.ConsoleWriteHeader("CSV Customer Segmentation");
            using (var w = new System.IO.StreamWriter(csvlocation))
            {
                w.WriteLine($"LastName,SelectedClusterId");
                w.Flush();
                predictions.ToList().ForEach(prediction => {
                    w.WriteLine($"{prediction.LastName},{prediction.SelectedClusterId}");
                    w.Flush();
                });
            }

            Console.WriteLine($"CSV location: {csvlocation}");
        }

        private static void SaveCustomerSegmentationPlotChart(IEnumerable<ClusteringPrediction> predictions, string plotLocation)
        {
            Common.ConsoleHelper.ConsoleWriteHeader("Plot Customer Segmentation");

            var plot = new PlotModel { Title = "Customer Segmentation", IsLegendVisible = true };

            var clusters = predictions.Select(p => p.SelectedClusterId).Distinct().OrderBy(x => x);

            foreach (var cluster in clusters)
            {
                var scatter = new ScatterSeries { MarkerType = MarkerType.Circle, MarkerStrokeThickness = 2, Title = $"Cluster: {cluster}", RenderInLegend=true };
                var series = predictions
                    .Where(p => p.SelectedClusterId == cluster)
                    .Select(p => new ScatterPoint(p.Location[0], p.Location[1])).ToArray();
                scatter.Points.AddRange(series);
                plot.Series.Add(scatter);
            }

            plot.DefaultColors = OxyPalettes.HueDistinct(plot.Series.Count).Colors;

            var exporter = new SvgExporter { Width = 600, Height = 400 };
            using (var fs = new System.IO.FileStream(plotLocation, System.IO.FileMode.Create))
            {
                exporter.Export(plot, fs);
            }

            Console.WriteLine($"Plot location: {plotLocation}");
        }

        private static void OpenChartInDefaultWindow(string plotLocation)
        {
            Console.WriteLine("Showing chart...");
            var p = new Process();
            p.StartInfo = new ProcessStartInfo(plotLocation)
            {
                UseShellExecute = true
            };
            p.Start();
        }
    }
}
